Frameworks/CoreGraphics/DWriteFontCollectionHelper.mm

```diff
@@ -0,0 +1,207 @@
+    // Use S_FALSE to represent 'not found'
+    RETURN_HR_IF_NULL(S_FALSE, fontCollection);
+
+    size_t fontFamilyIndex;
+    BOOL fontFamilyExists;
+
+    RETURN_IF_FAILED(fontCollection->FindFamilyName(unicharFamilyName, &fontFamilyIndex, &fontFamilyExists));
+    if (fontFamilyExists) {
+        return fontCollection->GetFontFamily(fontFamilyIndex, outFontFamily);
+    }
+
+    return S_FALSE;
+}
+
+HRESULT DWriteFontCollectionHelper::_GetFontListForFamilyName(CFStringRef familyName, _Outptr_ IDWriteFontList** outFontList) {
+    ComPtr<IDWriteFontCollection> fontCollection = _GetFontCollection();
+
+    // If no/empty font collection, fail to return a font list
+    // Use of E_FAIL rather than a more nuanced HRESULT is permissable due to this being a private function only used in
+    // _InitializePropertiesMap()
+    RETURN_HR_IF_NULL(E_FAIL, fontCollection);
+
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    size_t index = 0;
+    BOOL exists = FALSE;
+
+    const auto& unicharFamilyName = Strings::VectorFromCFString(familyName);
+    const wchar_t* data = reinterpret_cast<const wchar_t*>(unicharFamilyName.data());
+    RETURN_IF_FAILED(fontCollection->FindFamilyName(data, &index, &exists));
+
+    // Trying to create with a nonexistent font
+    RETURN_HR_IF(E_INVALIDARG, !exists);
+
+    ComPtr<IDWriteFontFamily> fontFamily;
+    RETURN_IF_FAILED(fontCollection->GetFontFamily(index, &fontFamily));
+
+    RETURN_IF_FAILED(
+        fontFamily->GetMatchingFonts(DWRITE_FONT_WEIGHT_NORMAL, DWRITE_FONT_STRETCH_NORMAL, DWRITE_FONT_STYLE_NORMAL, outFontList));
+
+    return S_OK;
+}
+
+void DWriteFontCollectionHelper::_InitializePropertiesMap() {
+    // Initialize m_propertiesMap
+    if (!m_propertiesMap) {
+        m_propertiesMap = std::make_shared<_DWriteFontPropertiesMap>();
+    }
+
+    woc::unique_cf<CFArrayRef> familyNames{ CopyFontFamilyNames() };
+
+    // Iterate over all the font families
+    CFIndex familyCount = CFArrayGetCount(familyNames.get());
+    for (size_t i = 0; i < familyCount; ++i) {
+        CFStringRef familyName = static_cast<CFStringRef>(CFArrayGetValueAtIndex(familyNames.get(), i));
+
+        ComPtr<IDWriteFontList> fontList;
+        if (FAILED(_GetFontListForFamilyName(familyName, &fontList))) {
+            continue;
+        }
+
+        // Iterate over each font in this family
+        size_t fontCount = fontList->GetFontCount();
+        for (size_t j = 0; j < fontCount; j++) {
+            ComPtr<IDWriteFont> font;
+            if (FAILED(fontList->GetFont(j, &font))) {
+                continue;
+            }
+
+            // For each font, fill out a _DWriteFontProperties
+            auto info = std::make_shared<_DWriteFontProperties>();
+            ComPtr<IDWriteLocalizedStrings> displayName;
+            ComPtr<IDWriteLocalizedStrings> postScriptName;
+            BOOL exist;
```

Comment:
[https://msdn.microsoft.com/en-us/library/windows/desktop/dd371147(v=vs.85).aspx](https://msdn.microsoft.com/en-us/library/windows/desktop/dd371147(v=vs.85).aspx)
exists [out]
Type: BOOL*
When this method returns, TRUE if the font contains the specified string ID; otherwise, FALSE.

I think it's probably safe to interpret that as saying 'this will always be populated'?

---
In reply to: [90931716](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90931716) [](ancestors = 90931716,90747010,90742971)